src/back/spirv/parser.rs

```diff
@@ -0,0 +1,1545 @@
+                        }
+                        64 => {
+                            let (low, high) = ((val >> 32) as u32, val as u32);
+                            instruction.add_operand(low);
+                            instruction.add_operand(high);
+                        }
+                        _ => unreachable!(),
+                    },
+                    _ => unreachable!(),
+                }
+
+                (instruction, id)
+            }
+            crate::ConstantInner::Float(val) => {
+                let type_id = self.get_type_id(arena, &constant.ty);
+
+                let mut instruction = Instruction::new(Op::Constant);
+                instruction.set_type(type_id);
+                instruction.set_result(id);
+
+                let ty = &ir_module.types[constant.ty];
+                match ty.inner {
+                    crate::TypeInner::Scalar { kind: _, width } => match width {
+                        32 => {
+                            instruction.add_operand((val as f32).to_bits());
+                        }
+                        64 => {
+                            let bits = f64::to_bits(val);
+                            let (low, high) = ((bits >> 32) as u32, bits as u32);
+                            instruction.add_operand(low);
+                            instruction.add_operand(high);
+                        }
+                        _ => unreachable!(),
+                    },
+                    _ => unreachable!(),
+                }
+
+                (instruction, id)
+            }
+            crate::ConstantInner::Bool(val) => {
+                let type_id = self.get_type_id(arena, &constant.ty);
+                let mut instruction;
+
+                if val {
+                    instruction = Instruction::new(Op::ConstantTrue);
+                    instruction.set_type(type_id);
+                    instruction.set_result(id);
+                } else {
+                    instruction = Instruction::new(Op::ConstantFalse);
+                    instruction.set_type(type_id);
+                    instruction.set_result(id);
+                }
+                (instruction, id)
+            }
+            crate::ConstantInner::Composite(ref constituents) => {
+                let type_id = self.get_type_id(arena, &constant.ty);
+
+                let mut instruction = Instruction::new(Op::ConstantComposite);
+                instruction.set_type(type_id);
+                instruction.set_result(id);
+
+                for constituent in constituents.iter() {
+                    let id = self.get_constant_id(constituent, &ir_module);
+                    instruction.add_operand(id);
+                }
+
+                (instruction, id)
+            }
+        }
+    }
+
+    fn get_pointer_id(
+        &mut self,
+        arena: &crate::Arena<crate::Type>,
+        handle: &crate::Handle<crate::Type>,
+        class: &StorageClass,
```

Comment:
all of these should be passed by value
only bigger than word sizes should be passed by reference